### PR TITLE
[GEOS-4613] - Expose more JVM statistics on the web gui

### PR DESCRIPTION
**Pull request by fostash**
[GEOS-4613](https://osgeo-org.atlassian.net/browse/GEOS-4613)

Add new tab `JVM Console` under the Server Status page with the possibility to see and download dumps for Threads and Heap
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/web/core/src/main/java/org/geoserver/web/system/status/ConsoleInfoUtils.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                            .map(processName -> Arrays.asList(processName.split("@")))
+                            .filter(elements -> !elements.isEmpty())
+                            .map(maybePid -> Integer.valueOf(maybePid.get(0)))
+                            .orElseThrow(
+                                    () ->
+                                            new IOException(
+                                                    "Problem on getting the PID for the java process"));
+
+            final Process jmapProcess =
+                    Runtime.getRuntime().exec(String.format("jmap -histo:live %d", pid));
```
Comment (mprins) [inline src/web/core/src/main/java/org/geoserver/web/system/status/ConsoleInfoUtils.java:51]: Oracle lists `jmap` as 
> Note:This command is experimental and unsupported.

on https://docs.oracle.com/en/java/javase/11/tools/jmap.html

So I'm not sure if it's a good idea to rely on it being available and/or accessible to the user geoserver is running from.

**Review by mprins — COMMENTED** [review]
I'd prefer this kind of thing to only be available to the ADMIN roles of GeoServer, I may be missing something, but it looks like any user could create dump files

**Review by aaime — COMMENTED** [review]
As a user-facing change, documentation needs to be added.

I've tried an interactive test too, got a failure:

```
Unable to find component with id 'dump.thread' in [JVMConsolePanel [Component id = panel]]
	Expected: 'tabs:panel:dump.thread'.
	Found with similar names: 'tabs:panel:dumpThread'
 MarkupStream: [markup = file:/home/aaime/devel/git-gs/src/web/core/target/classes/org/geoserver/web/system/status/JVMConsolePanel.html
<wicket:panel>
    <div style="width:95%">
        <div class="button-group selfclear">
            <a href="#" wicket:id="dump.thread"><wicket:message key="threadDump">Thread dump</wicket:message></a>
            <a href="#" wicket:id="dump.heap"><wicket:message key="heapDump">Heap dump</wicket:message></a>
        </div>

        <div><textarea readonly="readonly" style="width:100%;height:50em;font-size:small" wicket:id="dumpContent"></textarea></div>
        <a href="#" wicket:id="download"><wicket:message key="download">download as dump text</wicket:message></a>
    </div>
```

Indeed both button wicket ids are not matching what is used in the code.. easy fix needed.

The "Thread dump" button does what it's supposed to:

image

Wondering if we could also get a deadlock detection like it's done in jstack... but if not possible or not trivial, no worries, just thinking out loud here!

The "Heap dump" button is misleading, no actual heap dump is produced, but a "Heap histogram" is generated instead (I'd re-label the button):

image

The "downlload" button at the bottom of the UI is not functional. Then again, this PR is still in draft form, so I guess it is to be expected.

**Comment by aaime** [comment]
> I'd prefer this kind of thing to only be available to the ADMIN roles of GeoServer, I may be missing something, but it looks like any user could create dump files

The tab is in the "server status" page, which is AFAIK only reachable when logged in as an administrator... wondering if it shows up when logged in as a "workspace administrator" though (hopefully not?).

File: src/web/core/src/test/java/org/geoserver/web/system/status/JVMConsolePanelTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        login();
+        tester.getApplication().getMarkupSettings().setStripWicketTags(false);
+        tester.startPage(StatusPage.class);
+        tester.assertRenderedPage(StatusPage.class);
+        tester.clickLink("tabs:tabs-container:tabs:3:link", true);
+    }
+
+    @Test
+    public void testLoadConsoleTab() {
+        TagTester threadDumpBtn = tester.getTagByWicketId("dumpThread");
```
[the remark this answers is not in the document]
Comment (fostash) [reply src/web/core/src/test/java/org/geoserver/web/system/status/JVMConsolePanelTest.java:28]: Yes I was trying to write one more test to check the result on the textarea after click on button, but I didn't find a way to click on the button, `clickLink` doesn't find the element.

[the remark this answers is not in the document]
Comment (aaime) [reply src/web/core/src/test/java/org/geoserver/web/system/status/JVMConsolePanelTest.java:28]: Could it be due to the misaligned wicket ids reported above?

File: src/web/core/src/main/java/org/geoserver/web/system/status/ConsoleInfoUtils.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadInfo;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/** utility interface for thread and memory heap info */
+public interface ConsoleInfoUtils {
```
Comment (aaime) [inline src/web/core/src/main/java/org/geoserver/web/system/status/ConsoleInfoUtils.java:17]: Using an interface for utility methods is technically possible but quite weird (e.g. https://stackoverflow.com/questions/25098937/java-8-interface-with-static-methods-instead-of-static-util-class  or https://stackoverflow.com/questions/30905236/are-interfaces-a-valid-substitute-for-utility-classes-in-java-8).

I'd be more confortable with the classic final class with private constructor, and static imports in the place of usage.

**Comment by jodygarnett** [comment]
why was this merged (and then back ported without docs?)

**Comment by aaime** [comment]
Because I did not notice the lack of docs, sorry... now that I check it better it has been sitting open for a long time, plenty of time for others to review too. 
@fostash is no longer working with us, not sure who can write the docs now... are you interested?

**Pull request merged**